runtime/src/iree/base/internal/math.h

```diff
@@ -336,9 +336,60 @@ static inline float iree_math_make_f32_from_bits(uint32_t src, int exp_bits,
     // Zero. Leave f32_exp and f32_mantissa as zero.
   } else {
-    f32_exp = ((src_exp >> src_exp_shift) + f32_exp_bias - src_exp_bias)
-              << f32_exp_shift;
-    f32_mantissa = src_mantissa << (f32_mantissa_bits - src_mantissa_bits);
+    if (src_exp == 0) {
+      // Source is denormal, nonzero.
+
+      // Case when the source and destination types have the same exponent
+      // range (bfloat16/f32).
+      if (src_exp_bits == f32_exp_bits) {
+        f32_exp = src_exp << (f32_exp_shift - src_exp_shift);
+        f32_mantissa = src_mantissa << (f32_mantissa_bits - src_mantissa_bits);
+      } else {
+        // Source is denormal, destination is normal. Careful with the
+        // arithmetic exponents, as denormals are a special case:
+        // The source being denormal means that the arithmetic exponent is one
+        // plus what is normally encoded in the exponent bits. We won't need
+        // this, but we will refer to this in comments:
+        //     int src_arithmetic_exp = 1 - src_exp_bias;
+        // The f32 destination being normal means it doesn't have that
+        // adjustment by one.
+        int f32_arithmetic_exp = -src_exp_bias;
+        // Compute the final f32 exponent bits.
+        f32_exp = (f32_arithmetic_exp + f32_exp_bias) << f32_exp_shift;
+        // Now the hard part: compute the final f32 mantissa. Unlike the source,
+        // the f32 result is a normal value, which means that it is encoded as
+        //   (1 + mantissa_value) * 2^dst_arithmetic_exponent
+        // Where mantissa_value is a number in the half-open interval [0, 1).
+        // Scaling this to the actual integer encoded in the mantissa bits, that
+        // value is:
+        //    ((1 << f32_mantissa_bits) + f32_mantissa)
+        //                        * 2^(f32_arithmetic_exp - f32_mantissa_bits)
+        // Now we plug that into an equation of that value with the source
+        // value which is a denormal and hence does not have the corresponding
+        // "1 + " term. After some algebra isolating f32_mantissa on the left
+        // hand side of the equantion, and since per the above we have
+        //   -f32_arithmetic_exp + src_arithmetic_exp == 1
+        // we arrive at this:
+        int32_t unclamped_f32_mantissa =
+            (src_mantissa << (f32_mantissa_bits - src_mantissa_bits + 1)) -
+            (1 << f32_mantissa_bits);
+        // We don't want the f32_mantissa to run outside the range [0, 2^23 - 1]
+        // so we just clamp. I *think* that is correct. The reasoning is that
+        // first, there is no question of rounding here since we are expanding
+        // from a narrow to a wider type, and so the only question really is:
+        // did we pick the correct exponent above? As long as f32_exp is correct
+        // it forces a range of representable values and that means that
+        // clamping to the mantissa range is the correct rounding to do.
```

Comment:
Hold on - wouldn't the mantissa being off to the left mean you need to increase the exponent and shilt right?

That is, if there's a 2^23 bit set, shift right by 1 and increase the exponent by 1 to compensate?

Reply:
Deleted this entire code as part of the changes in the latest commit!